### ISSUE TITLE
Installing LSP unsuccessful. Error: 500  (Internal Server Error)

### ISSUE THREAD

**Issue by Juan-132** [krassowski/jupyterlab-lsp#123]
## Description
This seems like a great extension to me, but i can get it to install/work properly. :)

I installed it via conda, and all seems to resolve OK.
```
(base) C:\>conda install --insecure -c conda-forge python-language-server r-languageserver
Collecting package metadata (current_repodata.json): done
Solving environment: done
```

But when I try click 'install' within the extension pane in Jupyter Lab:
I first get the Server Companion warning: I click 'OK'. 
Then it gives me an Internal Server Error
![image](https://user-images.githubusercontent.com/59505296/71766493-a8663f00-2f00-11ea-9edf-c3aef2afb6a6.png)
I can't find anything useful in the documentation about this.

Conda 3.7
[conda list.txt](https://github.com/krassowski/jupyterlab-lsp/files/4021962/conda.list.txt)

**Comment by krassowski** [comment]
Did you install the python package (`jupyter-lsp`) first?

**Comment by krassowski** [comment]
Also, if you insist on installing from the package manager (and not using the command line instructions provided [here](https://github.com/krassowski/jupyterlab-lsp#installation)), please make sure that you re-start your JupyterLab after installing `jupyter-lsp` and before installing the `@krassowski/jupyterlab-lsp` extension from the Extension Manager.

**Comment by Juan-132** [comment]
Thank you for the quick reply krassowski.
I indeed forgot to run step 1 (my assumption in error was that as only 'pip' was mentioned, I could go on to the 'conda' install mentioned after step 3).

I just ran step 1 and hit has now been installed successfully:

```
(base) C:\>conda list jupyter-lsp
# packages in environment at C:\ProgramData\Anaconda3:
#
# Name                    Version                   Build  Channel
jupyter-lsp               0.6.0b0                  pypi_0    pypi
```

I also tried to run step 2 from the commandline, but got attached error
[error step 2.txt](https://github.com/krassowski/jupyterlab-lsp/files/4022099/error.step.2.txt)
It seems to fail on a version check for nodejs, This is strange, as it has indeed been installed:

```
(base) C:\>conda list nodejs
# packages in environment at C:\ProgramData\Anaconda3:
#
# Name                    Version                   Build  Channel
nodejs                    0.12.18                       0    tejzpr
```

I also restarted the jupyter lab server, and tried to install the extension again from within Jupyter Lab. This still gives me the Error: 500.


Would you have any other suggestions as to why installation of the extension (step 2) fails?

**Comment by bollwyvl** [comment]
Needs node 8 or newer:

 `conda install -c conda-forge nodejs`

But, I highly recommend:
- use a conda env
- adapt what you need from the environment.yml in the repo, which is what
we test/demo from.

 installing stuff in any unpredictable order with pip/conda/conda-forge can
definitely tools something that won't ever work again, especially on
windows.

**Comment by Juan-132** [comment]
Hi bollwyvl, I now installed the newest version of nodejs (13).
But that does not seem to work due to a conflict with npm. 

```
(base) C:\>conda list nodejs
# packages in environment at C:\ProgramData\Anaconda3:
#
# Name                    Version                   Build  Channel
nodejs                    13.0.0                        1    conda-forge

(base) C:\>jupyter labextension install @krassowski/jupyterlab-lsp
An error occured.
ValueError: "@krassowski/jupyterlab-lsp" is not a valid npm package
See the log file for details:  C:\Users\XXXXXX\AppData\Local\Temp\jupyterlab-debug-tcgdyvs1.log
[error nodejs.txt](https://github.com/krassowski/jupyterlab-lsp/files/4022214/error.nodejs.txt)
```

**Comment by Juan-132** [comment]
An observation from my point of view:
The steps to install this extension for Jupyter Lab are fairly complex, even more so if I compare how easy it has been today installing all other packages to setup my environment. The reason I believe it's complex, is that the installation instruction requires you to install everything step-by-step. This is especially cumbersome with extra dependencies/version incompatability like NPM and NoteJS above.

I believe this is even more so difficult for novice users, and those more geared towards data science (like myself); instead of those with focused on developing software/making software work.

I believe we can improve the project if we more the installation process simpler.
If I may make a suggestion: Could we create a `'conda install jupyter-lsp-python'` command; which would install everything to make Jupyter language support in Jupyter Lab work.

The conda command would install/check all the dependencies needed to make it work:
* Install/Check for supported versions of npm/nodejs
* Install server extension
* Install Jupyter Lab extension
* Install Python Language Server

**Comment by bollwyvl** [comment]
We appreciate all the efforts of prospective users to boldly try and get
stuff up and running.

This is still very beta, open source software, being built by two busy
folks in their spare time to the best of their abilities.

I personally agree that offering a conda package will end up being the best
recommended installation option, in line with the Jupyter recommendation of
using anaconda. I'm itching to get this into a conda-forge recipe, but the
robustness isn't there yet for me to propose it, while running another
conda channel is seldom the right answer to these kinds of problems. But of
course some people just don't like conda, so we'll have to have _n_ paths
that work without it.

Compounding these issues are that language servers are as complicated as
jupyter kernels, sometimes more so, and implement the specification to
various degrees of success, sometimes differently on different platforms
and with different limitations. We've had some success getting the python
and r servers packaged pretty well on cf, but the node-based stuff is, and
will remain, kind of a nightmare: this is why vscode/atom/etc run their
own, separate registries, and install their own node. We're not going to be
doing that.

**Comment by Juan-132** [comment]
Thanks for your reply.
I can understand the complexity, and that this might not be the right time to implement this yet. I really value the effort you guys put in, also in you spare time. And hope this can be realized at some point in the future. Cheers!

**Comment by krassowski** [comment]
Thanks for the discussion. @Juan-132 I took a look at our documentation, and:
- we already have an entry on the nodejs version
- we do not have an explicit text on the necessity of installing the server extension, but I do not see how to make it more clear than it is currently without adding unnecessarily long explanations (as you may have experienced the instructions are quite long already and ppl tend to skip long instructions)
- we did not mention the need to restart JupyterLab nor that the installation via ExtensionManager requires a specific order + a restart; I tried to fix that in #126 - can I ask you to have a look on this PR and let me know if this would have helped you/how it could be improved?

Also, if you enjoy the extension, please feel welcome to propose other documentation improvements and otherwise participate in the creation of this add-on - any help is appreciated!

**Comment by Juan-132** [comment]
Sure.
As you correctly state, this instruction is getting quite long right now. Therefore I would choose to keep it more instructional/ like a recipe :). Below I included a draft, to replace everything from the section 'Prerequisites' and down.
Please feel free to take from this what you find useful.


## Python Prerequisite

This extension requires Python 3.5 or higher. Please upgrade before installing LSP.

## Step-by-step Installation Guide

To install LSP support you need to follow this 5-step installation guide:

**Step 1.** Make sure your JupyterLab is up-to-date: `pip install --upgrade jupyterlab`
> Note: LSP requires JupyterLab version 1.2.3 or higher.

**Step 2.** Install LSP support for Jupyter server: `pip install --pre jupyter-lsp`
**Step 3.** Install LSP support for the actual Jupyter Lab notebook: `jupyter labextension install @krassowski/jupyterlab-lsp`

> Note: To install LSP for the notebook, you need to have NodeJS installed.
> To install: `pip install nodejs`. Please make sure you have version 8, 9, 10, 11 or 12 installed.

**Step 4.** Add Python and R language server support to LSP. You can do this via PIP or Conda:

   ```bash
   *Option 1 (via PIP):*
   pip install python-language-server[all] 
   R -e 'install.packages("languageserver")'

   *Option 2 (via Conda):*
   conda install -c conda-forge python-language-server r-languageserver
   ```
**Step 5.** Done. Restart Jupyter Lab if it was running, and enjoy the magic :)


## Adding support for your programming language

It's easy to add support for your language of choice to LSP. Please see our full list of [supported language servers](./LANGUAGESERVERS.md). This page includes installation commands for a wide range of languages such as HTML, Bash or javascript.
In general, any LSP server from the [Microsoft list](https://microsoft.github.io/language-server-protocol/implementors/servers/) should work after [some additional configuration](./CONTRIBUTING.md#specs).


## Enable opening files outside root (Optional, Linux/OSX-only)
To enable opening files outside of the root directory (the place where you start JupyterLab), create `.lsp_symlink` and symlink your `/home`, or any other location which includes the files that you
wish to make possible to open in there:

```bash
mkdir .lsp_symlink
cd .lsp_symlink
ln -s /home home
```

If your user does not have sufficient permissions to traverse the entire path, you will not be able to open the file. A more detailed guide on symlinking
(written for a related jupyterlab-go-to-definition extension) is available [here](https://github.com/krassowski/jupyterlab-go-to-definition/blob/master/README.md#which-directories-to-symlink).

**Comment by krassowski** [comment]
Thank you! I will take inspiration from what you proposed when working on a re-wrire before the next major (0.8) release. I like the idea of adding "step" indicators so that users do not try to jump around. Also, I think that we could have a short version in the readme, and a longer one in an article/docs.

In the meantime, I will close this issue as the original problem [was solved](https://github.com/krassowski/jupyterlab-lsp/issues/123#issuecomment-570803804), and the documentation improvements are already tracked elsewhere (in #126).
